### PR TITLE
feat: Expose data provider setter and getters in components

### PR DESCRIPTION
**Pull request by mshabarov**
## Description

Some components, like Grid and ComboBox, have publicly exposed `setDataProvider` and `getDataProvider` methods - others do not, and they were modified recently to have these methods private. These two methods are made public in this patch, mainly for API consistency and for letting developers use filterable data providers (non Void filter generic parameter), which is not possible with `setItems` methods in most of the components.

Motivated by a comment from user https://github.com/vaadin/flow/issues/15665#issuecomment-1527067023 and internal discussion about having public `getDataProvider` method in all components.

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [ ] The issue is created in the corresponding repository and I have referenced it.
- [ ] I have added tests to ensure my change is effective and works as intended.
- [ ] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.
- [x] I have not completed some of the steps above and my pull request can be closed immediately.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java [hunk 1]
```diff
@@ -320,7 +320,18 @@ public T getItem() {
         }
     }
 
-    private void setDataProvider(DataProvider<T, ?> dataProvider) {
+    /**
+     * {@inheritDoc}
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:324]: I'll check and fix, thanks!

File: vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java [hunk 2]
```diff
@@ -320,7 +320,18 @@ public T getItem() {
...
+     *
+     * Use instead one of the {@code setItems} methods which provide access to
+     * either {@link CheckboxGroupListDataView} or
+     * {@link CheckboxGroupDataView}.
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:328]: Makes sense, will remove this paragraph.

File: vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBoxBase.java
[the remark this answers is not in the document]
Comment (mshabarov) [reply vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBoxBase.java:1]: For `ComboBox` all or at least the majority of removed `setDataProvider` methods can be 1:1 replaced by existing `setItems`. But I will double check it.

[the remark this answers is not in the document]
Comment (mshabarov) [reply vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBoxBase.java:1]: Added two methods again. They don't have a direct replacement. Others do have.

File: vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mshabarov) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java]: Done.

**Review by sissbruecker — APPROVED** [review]
Thanks for the updates, LGTM

**Pull request merged**